### PR TITLE
[FIXED JENKINS-41311] Fix NPE in CleanupJobsOnDelete.onDeleted

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-41311](https://issues.jenkins-ci.org/browse/JENKINS-41311)

Still needs a test but I'm not sure how to actually reproduce this yet. Suggestions?

cc @reviewbybees esp @amuniz @batmat

### CODE REVIEW

**Comment by jglick** [comment]
Reported in #6 but without steps to reproduce.

BTW mark both `getMilestonesByOrdinalByJob()` methods `@Nonnull`.

**Comment by jglick** [comment]
Start by writing a `@LocalData` test with a home dir written with the previous version of the plugin.

**Comment by jglick** [comment]
I think I got it. f197af9d3166d06d56bbbd15aad286bcb96bf1f6 was wrong. Read the Javadoc for `Descriptor.load`. That should be enough to get a reproducible test case. In fact I suspect you do not even need `@LocalData` for this purpose—`cleaupTrackedJobsOnDeleted` (the only test coverage for this hook) runs a build which uses `milestone` *before* deleting the job, thus `load`ing the descriptor via `MilestoneStepExecution.load`. Try writing a test case that does nothing but, say,

```java
r.createFreestyleProject().delete();
```

and I bet you will get the error!

Delete `MilestoneStepExecution.load`. Not only is it not getting called often enough to prevent this NPE, it is getting called *too much*—only needed once at startup.

**Comment by abayer** [comment]
Bingo, got it - thanks for the pointers, @jglick!

File: src/test/java/org/jenkinsci/plugins/pipeline/milestone/CleaupJobsTest.java [hunk 1]
```diff
@@ -48,4 +57,22 @@ public void cleaupTrackedJobsOnDeleted() throws Exception {
...
+        LogRecorderManager mgr = j.jenkins.getLog();
+        LogRecorder.Target t = new LogRecorder.Target(ItemListener.class.getName(), Level.ALL);
+        recorder.targets.add(t);
+        recorder.save();
+        t.enable();
+        mgr.logRecorders.put("recorder", recorder);
+
+        j.createFreeStyleProject().delete();
+        for (LogRecord r : recorder.getLogRecords()) {
+            assertFalse(r.getMessage().contains("failed to send event to listener of class org.jenkinsci.plugins.pipeline.milestone.MilestoneStepExecution$CleanupJobsOnDelete"));
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/pipeline/milestone/CleaupJobsTest.java]: Ah, right—`ItemListener` catches this.

File: src/test/java/org/jenkinsci/plugins/pipeline/milestone/CleaupJobsTest.java [hunk 2]
```diff
@@ -48,4 +57,22 @@ public void cleaupTrackedJobsOnDeleted() throws Exception {
...
+    @Issue("JENKINS-41311")
+    @Test
+    public void noNPEOnCleanup() throws Exception {
+        LogRecorder recorder = new LogRecorder("recorder");
+        LogRecorderManager mgr = j.jenkins.getLog();
+        LogRecorder.Target t = new LogRecorder.Target(ItemListener.class.getName(), Level.ALL);
+        recorder.targets.add(t);
+        recorder.save();
+        t.enable();
+        mgr.logRecorders.put("recorder", recorder);
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/pipeline/milestone/CleaupJobsTest.java]: Use `LoggerRule` and save yourself a lot of effort.

**Comment by abayer** [comment]
@reviewbybees done

**Pull request merged**